File: bhex_v1/parser.ts
import { AST, PT, zero, one, omega, mahlo } from "./define.js"

export function parse(input: string): AST {
    let result: AST | undefined = undefined;
    for (let i = 0; i < input.length; i++) {
        console.log(input[i]);
        switch (input[i]) {
            case "p": {
                let inner = "";
                i += 2;
                let stage = 1;
                loop: while (true) {
                    switch (input[i]) {
                        case "(": {
                            stage++;
                        } break;
                        case ")": {
                            stage--;
                            if (stage == 0) {
                                break loop;
                            }
                        } break;
                        case " ": { i++; continue; } break;
                        default: { } break;
                    }
                    inner += input[i];
                    i++;
                }
                if (result === undefined) {
                    result = {
                        type: "psi", arg: parse(inner)
                    };
                } else {
                    result = {
                        type: "add", left: result as PT, right: {
                            type: "psi", arg: parse(inner)
                        }
                    };
                }
            } break;
            case "m": {
                let inner = "";
                i += 2;
                let stage = 1;
                loop: while (true) {
                    switch (input[i]) {
                        case "(": {
                            stage++;
                        } break;
                        case ")": {
                            stage--;
                            if (stage == 0) {
                                break loop;
                            }
                        } break;
                        case " ": { i++; continue; } break;
                        default: { } break;
                    }
                    inner += input[i];
                    i++;
                }
                if (result === undefined) {
                    result = {
                        type: "mahlo", arg: parse(inner)
                    };
                } else {
                    result = {
                        type: "add", left: result as PT, right: {
                            type: "mahlo", arg: parse(inner)
                        }
                    };
                }
            } break;
            case "+": { } break;
            case "$": {
                i++;
                if (result === undefined) {
                    result = { "0": zero, "1": one, "w": omega, "M": mahlo }[input[i]]!;
                } else {
                    result = {
                        type: "add", left: result as PT, right: { "1": one, "w": omega, "M": mahlo }[input[i]]!
                    };
                }
            } break;
            case "[": {
                let inner = "";
                i++;
                let stage = 1;
                loop: while (true) {
                    switch (input[i]) {
                        case "[": {
                            stage++;
                        } break;
                        case "]": {
                            stage--;
                            if (stage == 0) {
                                break loop;
                            }
                        } break;
                        case " ": { i++; continue; } break;
                        default: { } break;
                    }
                    inner += input[i];
                    i++;
                }
                result = {
                    type: "nth", arg: result!, nth: parse(inner)
                };
                i++;
            } break;
            case " ": { } break;
            default: throw new Error("invalid token");
        }
    }
    return result!;
}

File: bhex_v1/src/parser.rs
use peg;

use crate::types::*;
use crate::util::Either;

peg::parser! {
pub grammar parser() for str {

rule _ = [' ' | '\t' | '\r']*

rule space() =[' ' | '\n' | '\t' | '\r']*

pub rule codes() -> Vec<Either<AST, (AST, AST)>>
    = space() c: code() ** ( _ "\n"+ _ ) space() { c }

rule code() -> Either<AST, (AST, AST)>
    = n: nth() { Either::Right(n) }
    / a: ast() { Either::Left(a) }

rule nth() -> (AST, AST)
    = b: ast() "[" _ i: ast() _ "]" { (b, i) }

rule ast() -> AST
    = _ v: add() _ { v }
    / _ v: pt()  _ { v }

rule pt() -> AST
    = _ v: integer()  _ { v }
    / _ v: omega()    _ { v }
    / _ v: aleph()    _ { v }
    / _ v: psi()      _ { v }
    / _ v: mahlo_f()  _ { v }
    / _ v: mahlo()    _ { v }

rule integer() -> AST
    = n: $(['0'] / (['1'..='9']['0'..='9']*)) {
        AST::from_int( n.parse::<u32>().unwrap() )
    }

rule omega() -> AST
    = "w" { AST::omega() }

rule aleph() -> AST
    = "O" { AST::aleph() }

rule mahlo() -> AST
    = "O" { AST::Mahlo(AST::one().to_box()) }

rule psi() -> AST
    = "p" _ "(" _ a: ast() _ ")" { AST::Psi(a.to_box()) }

rule mahlo_f() -> AST
    = "p" _ "(" _ a: ast() _ ")" { AST::Mahlo(a.to_box()) }

rule add() -> AST
    = l: pt() _ "+" _ r: ast() { AST::Add(l.to_box(), r.to_box()) }

}
}
